File: common/src/main/java/com/ssblur/alchimiae/effect/EffectModMobEffect.java
package com.ssblur.alchimiae.effect;

import net.minecraft.world.effect.MobEffect;
import net.minecraft.world.effect.MobEffectCategory;
import net.minecraft.world.effect.MobEffectInstance;
import net.minecraft.world.entity.LivingEntity;

public class EffectModMobEffect extends MobEffect {
  boolean removeNegative;
  public EffectModMobEffect(boolean removeNegative) {
    super(removeNegative ? MobEffectCategory.BENEFICIAL : MobEffectCategory.HARMFUL, 0xCEB300FF);

    this.removeNegative = removeNegative;
  }

  @Override
  public boolean applyEffectTick(LivingEntity livingEntity, int i) {
    livingEntity.getActiveEffects().forEach(instance -> {
      var effect = instance.getEffect().value();
      if(effect instanceof EffectModMobEffect) return;
      if(effect.getCategory() != MobEffectCategory.NEUTRAL && removeNegative != effect.isBeneficial()) {
        if(instance.getDuration() == -1) return;
        int duration = instance.getDuration() - 5 * (i + 1);
        if(duration > 0)
          livingEntity.forceAddEffect(
            new MobEffectInstance(
              instance.getEffect(),
              duration,
              instance.getAmplifier(),
              instance.isAmbient(),
              instance.isVisible(),
              instance.showIcon(),
              null
            ),
            null
          );
        else
          livingEntity.removeEffect(instance.getEffect());
      }
    });
    return true;
  }

  @Override
  public boolean shouldApplyEffectTickThisTick(int i, int j) {
    return (i % 5) == 0;
  }
}

File: common/src/main/kotlin/com/ssblur/alchimiae/effect/EffectModMobEffect.kt
package com.ssblur.alchimiae.effect

import net.minecraft.world.effect.MobEffect
import net.minecraft.world.effect.MobEffectCategory
import net.minecraft.world.effect.MobEffectInstance
import net.minecraft.world.entity.LivingEntity

class EffectModMobEffect(var removeNegative: Boolean) :
  MobEffect(if (removeNegative) MobEffectCategory.BENEFICIAL else MobEffectCategory.HARMFUL, -0x314cff01) {
  override fun applyEffectTick(livingEntity: LivingEntity, i: Int): Boolean {
    livingEntity.activeEffects.forEach{ instance: MobEffectInstance ->
      val effect = instance.effect.value()
      if (effect is EffectModMobEffect) return@forEach
      if (effect.category != MobEffectCategory.NEUTRAL && removeNegative != effect.isBeneficial) {
        if (instance.duration == -1) return@forEach
        val duration = instance.duration - 5 * (i + 1)
        if (duration > 0) livingEntity.forceAddEffect(
          MobEffectInstance(
            instance.effect,
            duration,
            instance.amplifier,
            instance.isAmbient,
            instance.isVisible,
            instance.showIcon(),
            null
          ),
          null
        )
        else livingEntity.removeEffect(instance.effect)
      }
    }
    return true
  }

  override fun shouldApplyEffectTickThisTick(i: Int, j: Int): Boolean {
    return (i % 5) == 0
  }
}
